ui/plugin-system/src/model/variables.ts

```diff
@@ -12,19 +12,23 @@
 // limitations under the License.
 
 import { ListVariableDefinition } from '@perses-dev/core';
-import { LegacyDatasources } from '../runtime';
+import { LegacyDatasources, VariableStateMap } from '../runtime';
 
 export type VariableOption = { label: string; value: string };
 
 export interface GetVariableOptionsContext {
   datasources: LegacyDatasources;
+  variables: VariableStateMap;
 }
 
 /**
  * Plugin for handling custom VariableDefinitions.
  */
 export interface VariablePlugin<Spec = unknown> {
   getVariableOptions: GetVariableOptions<Spec>;
+
+  /** Returns a list of variables name this variable depends on. Used to optimize fetching */
+  dependsOn?: (definition: ListVariableDefinition<Spec>) => string[];
```

Comment:
This is a new feature on a plugin that I think it worth discussing. Basically makes its easy for plugins to still "own" how template variables are consumed in the plugin but have the framework still understand the dependencies.

Reply:
> > Super cool demo. Need to think on the dependencies thing a bit, so more thoughts coming soon.
> 
> Thanks! The other approach I thought about was exposing a function called `queryKey` that gets passed into the [queryKey argument](https://tanstack.com/query/v4/docs/guides/query-keys) for react-query which would give you even more control.
> 
> For example:
> 
> ```ts
> export const PrometheusLabelValuesVariable: VariablePlugin<PrometheusLabelValuesVariableOptions> = {
>   getVariableOptions: async (definition, ctx) => {
>     //... implementation
>   },
>   queryKey: (definition, ctx) => { 
>     // return some unique string here that can consume variables, time range state, etc
>     return someKey;
>   }
> };
> ```
> 
> I ultimately decided against it because I felt it was too primitive and doesn't really solve the problem of the framework knowing about dependencies.

Yeah, at first glance for some reason, I thought you'd put this into the JSON spec and I was thinking to myself last night, "why isn't it just a function on the plugin". Then I came back and looked at the PR again and was like, "Oh, that's totally what I was thinking you should do". 😂

Reply:
My only question here is: do you need the entire definition as an argument, or just the `Spec`?

Reply:
Yeah, I've been thinking about changing up the other plugins too so they don't have access to the whole definition, just the spec (seems like it's not necessary anywhere I can think of), so if you wanted to do it, cool with me.